cs3api4lab/locks/metadata.py

```diff
@@ -0,0 +1,69 @@
+import json
+import time
+import datetime
+import urllib.parse
+
+from cs3api4lab.api.cs3_user_api import Cs3UserApi
+from cs3api4lab.locks.base import LockBase
+from cs3api4lab.exception.exceptions import FileLockedError
+
+
+class Metadata(LockBase):
+
+    def __init__(self, log, config):
+        super().__init__(log, config)
+        self.log = log
+        self.config = config
+        self.user_api = Cs3UserApi(log)
+        self.locks_expiration_time = self.config.locks_expiration_time
+
+    def handle_locks(self, file_path, endpoint):
+        lock = self._get_lock(file_path, endpoint)
+        if not lock:
+            self._lock_file(file_path, endpoint)
+            return
+        else:
+            if self._is_lock_mine(lock):
+                self._lock_file(file_path, endpoint)
+                return
+            if self._is_lock_expired(lock):
+                self._lock_file(file_path, endpoint)
+                return
+        raise FileLockedError("File %s is locked" % file_path)
+
+    def _generate_lock_entry(self):
+        user = self.get_current_user()
+        return {self.lock_name: urllib.parse.quote(json.dumps({
+            "username": user.username,
```

Comment:
Hi! This is our custom implementation of locks which uses file's arbitrary metadata entries. The CS3 implementation present in cs3.py does not include user's info manually.

Reply:
OK, but then how is the `Cs3._set_lock` method going to get the `lock_name` content? I'm probably missing something in the inheritance, but it seems the `Cs3` class implementation relies on this inherited function to actually create the payload.

Reply:
_lock_name_ is defined in the base class, and this will be used as _lock_id_ in case of cs3 implementation, in case of metadata implementation we need to manually include user's info into the entry

Reply:
Right, so my original comment holds: the `lock_id` is going to include details that are redundant compared to the metadata implementation. And also, when returning the lock in https://github.com/sciencemesh/cs3api4lab/pull/209/files#diff-828df65314d9ed749fee75882a0cc9a64a2bea77bbc9493a4b37a9c667b2f9bfR44 as the whole lock is returned, not just `lock_id`, the dict you obtain is not equivalent to the content you'd get from the pure metadata implementation.

My suggestion here would be to actually NOT try to be compatible and maintain both implementation, but rather adopt the native locks as they are and simplify the whole logic (no need to check for expired locks, etc.). Or if really you want to keep the two implementation, then `Cs3._get_lock` should return `lock.lock_id`, but the code looks a bit cluttered with redundant logic.

Reply:
I already corrected checking expiration time in _Cs3_ unnecessarily, do you see another overcomplications in _Cs3_ implementation?

And maybe I didn't highlighted it, but the methods like _set/_getlock etc though named the same, are not a part of the interface defined in base class, and in each case the locks are generated differently, though they share a common lock_id. So in case of _Metadata_ leaving only _created_ and _updated_ will make impossible to determine the lock owner. 

Wether we need to maintain 2 implementations it's up to @diocas, but I guess until cs3 locks are not implemented for localfs as well this is better than not having locks locally at all...

Reply:
I see, indeed it makes sense for now to keep both implementations until we're sure that all reasonably usable/used storage providers implement the lock API.

I'll have another look at the current code.